### PR TITLE
Packed transpose.

### PR DESCRIPTION
**Pull request by astojilj**
This PR is a split from PR #1448 and implements packed transpose
operation. Together with packed slice and packed pad, this completes
packed batchToSpace and spaceToBatch operations.

Compiled shader's userCode for transpose tensor of shape [3, 1001, 1] with perm=[0, 2, 1] is:

```
    void main() {
      ivec3 rc = getOutputCoords();
      vec4 result = vec4(0.);
      result[0] = getChannel(getA(rc.x,rc.z,rc.y), vec2(rc.z,rc.y));
      if(++rc.z < 1001) {
        result[1] = getChannel(getA(rc.x,rc.z,rc.y), vec2(rc.z,rc.y));
      }
      --rc.z;
      if(++rc.y < 1) {
        result[2] = getChannel(getA(rc.x,rc.z,rc.y), vec2(rc.z,rc.y));
        if(++rc.z < 1001) {
          result[3] = getChannel(getA(rc.x,rc.z,rc.y), vec2(rc.z,rc.y));
        }
      }  
      setOutput(result);
    }
```

[Benchmark](https://gist.github.com/astojilj/971be3037c115a9fa5fbb5d204d4da9e) on discrete GPU on [ASUS ROG-GL702VM](https://www.asus.com/us/ROG-Republic-Of-Gamers/ROG-GL702VM/) show performance improvement:

|          transpose 6D tensor of shape            | ** non packed ** | ** packed ** |
|:--------------------:|:-------------------------:|:------------------------:|
|       [1, 19, 4, 19, 4, 1]       |     ~5.2ms                      |         ~5.2ms                 |
| [1, 19, 4, 19, 4, 96] |           ~23ms                |             ~12ms             |
| [1, 19, 4, 19, 4, 960] |           ~152ms                |             ~56ms             |

Note: [1, 19, 4, 19, 4, 960] transpose happens in DeepLabV3+ MobileNetV2.

PERF

---
This change is [Reviewable](https://reviewable.io/reviews/tensorflow/tfjs-core/1504)

### CODE REVIEW

File: src/kernels/webgl/transpose_packed_gpu.ts [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+       if(${outputOrder[this.rank - 2]} < ${outputShape[this.rank - 2]}) {`,
+      `  ${outputOrder[this.rank - 1]} += 1;
+         if(${cLimit}) {`
+    ];
+    for (let i = 0; i < 4; i++) {
+      main += `
+        ${componentSetup[i]}
+          result[${i}] =
+            getChannel(getA(${switchedOrder.join()}), ${innerDims});
+      `;
```
Comment (annxingyuan) [inline src/kernels/webgl/transpose_packed_gpu.ts]: This is a bit hard to read - would it make sense to just write the iteration through the rgba channels out?

**Comment by astojilj** [comment]
@annxingyuan 
>      Run a benchmark on repeatedly transposing a large tensor 

Should the benchmark compare "pack + packed transpose + unpack" vs unpack transpose?

So far, performance improvement is verified with batchToSpace and spaceToBatch packed ([slice, pad and transpose](https://github.com/tensorflow/tfjs-core/blob/2e3142bdfd40173d4dcc125a5dfeeda7c8670aa7/src/kernels/backend_webgl.ts#L953) all packed make batch<->space packed).

If "pack + packed transpose + unpack" doesn't perform faster than unpacked transpose, we could make the choice, to use packed ot to use unpacked version, dependent on transpose operation input + environment flag?

Or you'd like to verify something else about packed transpose.

>     * Paste the compiled shader code into the description for this PR
Of course. I plan to rewrite the shader - expand it to make it easier to read.

**Comment by annxingyuan** [comment]
@astojilj 

I would say that as long as the packed transpose by itself is clearly faster than unpacked transpose, we can consider this an improvement (this is the standard we've been using for introducing other packed ops). Unlike with other ops we've packed (binary ops, unary ops, batchnorm, etc.), the additional work per fragment shader in packed transpose can't be vectorized, so we're essentially de-parallelizing work. Performance should still improve since the unpacked transpose work is well within the startup cost of a shader, but it would be good to have some measurements.

> So far, performance improvement is verified with batchToSpace and spaceToBatch packed (slice, pad and transpose all packed make batch<->space packed).

Are you saying you ran a benchmark on batchToSpace and saw an improvement? Could you include those measurements in this PR?

> If "pack + packed transpose + unpack" doesn't perform faster than unpacked transpose, we could make the choice, to use packed ot to use unpacked version, dependent on transpose operation input + environment flag?

Yes it might be a good idea to introduce heuristics like this at some point - let's not worry about it for now though :)

File: src/kernels/webgl/transpose_packed_gpu.ts (hunk 1 above)
Comment (astojilj) [reply src/kernels/webgl/transpose_packed_gpu.ts]: Done.

**Comment by astojilj** [comment]
@annxingyuan 
Here is a [live benchmark](https://01org.github.io/depth-camera-web-demo/nn/using-deeplab/index.html) of DeepLabV3+ MobileNetV2; use checkbox "Use packed array ops" to toggle environment flag WEBGL_PACK_ARRAY_OPERATIONS.

Setting WEBGL_PACK_ARRAY_OPERATIONS activates packed slice, transpose and pad, and inherently, batchToSpaceND and spaceToBatchND are packed, too.

Feel free to run it on different hardware - I've checked on [MacBook Pro (Retina, 15-inch, Mid 2014) with integrated card](https://support.apple.com/kb/sp704?locale=en_US): when all checkboxes are checked (also WEBGL_PACK_ARRAY_OPERATIONS), inference takes ~125ms.
When unchecking  "Use packed array ops" (WEBGL_PACK_ARRAY_OPERATIONS=false), inference takes ~150ms..

>https://github.com/tensorflow/tfjs-core/tree/master/integration_tests/benchmarks - see https://github.com/tensorflow/tfjs-core/blob/master/integration_tests/benchmarks/matmul_benchmarks.ts for an example) and compare to the unpacked version

I still need to check this.

**Comment by annxingyuan** [comment]
Thanks @astojilj
